### PR TITLE
SONARJAVA-1437 Precise issue location for a bunch of rules

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
 
   void addIssue(Tree tree, JavaCheck check, String message);
 
+  @Deprecated
   void addIssue(Tree tree, JavaCheck check, String message, @Nullable Double cost);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java [hunk 2]
```diff
@@ -76,7 +76,7 @@ public void visitCaseGroup(CaseGroupTree tree) {
     super.visitCaseGroup(tree);
 
     if (invalidCaseGroups.remove(tree)) {
-      context.addIssue(Iterables.getLast(tree.labels()), this, "End this switch case with an unconditional break, return or throw statement.");
+      context.reportIssue(this, Iterables.getLast(tree.labels()), "End this switch case with an unconditional break, return or throw statement.");
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java]: @benzonico I would raise the issue on `Iterables.getLast(tree.labels()).caseOrDefaultKeyword()`.

File: java-checks/src/main/java/org/sonar/java/checks/ThrowsFromFinallyCheck.java [hunk 3]
```diff
@@ -67,7 +67,7 @@ public void visitTryStatement(TryStatementTree tree) {
   @Override
   public void visitThrowStatement(ThrowStatementTree tree) {
     if(isInFinally() && !isInMethodWithinFinally){
-      context.addIssue(tree, this, "Refactor this code to not throw exceptions in finally blocks.");
+      context.reportIssue(this, tree, "Refactor this code to not throw exceptions in finally blocks.");
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/ThrowsFromFinallyCheck.java]: @benzonico what about `tree.throwKeyword()` only?

File: java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java [hunk 4]
```diff
@@ -37,6 +37,7 @@
 
   void addIssue(Tree tree, JavaCheck check, String message);
 
+  @Deprecated
```
Comment (mpaladin) [inline java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java]: @benzonico we should create a ticket to deprecate `addIssue` methods.

File: java-checks/src/main/java/org/sonar/java/checks/BooleanInversionCheck.java [hunk 5]
```diff
@@ -66,7 +66,7 @@ public void visitNode(Tree tree) {
         Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO,
         Tree.Kind.LESS_THAN, Tree.Kind.GREATER_THAN,
         Tree.Kind.LESS_THAN_OR_EQUAL_TO, Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
-      context.addIssue(tree, this, "Use the opposite operator (\"" + OPERATORS.get(((BinaryExpressionTree) expression).operatorToken().text()) + "\") instead.");
+      context.reportIssue(this, expression, "Use the opposite operator (\"" + OPERATORS.get(((BinaryExpressionTree) expression).operatorToken().text()) + "\") instead.");
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/BooleanInversionCheck.java]: @benzonico may be it is better to highlight `tree` which includes the `!`.

Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/BooleanInversionCheck.java]: Debatable but why not.

File: java-checks/src/main/java/org/sonar/java/checks/ThrowsFromFinallyCheck.java (hunk 3 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ThrowsFromFinallyCheck.java]: I disagree, I think it is worth to hilight the full statement here.

File: java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java (hunk 2 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java]: I disagree, we want to highlight the full case here IMO

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java (hunk 4 above)
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java]: https://jira.sonarsource.com/browse/SONARJAVA-1439

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java [hunk 6]
```diff
@@ -35,8 +35,16 @@
...
+   * @deprecated use reportIssue instead to benefit from precise issue location.
+   */
+  @Deprecated
   void addIssue(Tree tree, JavaCheck check, String message);
 
+  /**
+   * @deprecated use reportIssue instead to benefit from precise issue location.
+   */
+  @Deprecated
   void addIssue(Tree tree, JavaCheck check, String message, @Nullable Double cost);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java [hunk 7]
```diff
@@ -35,8 +35,16 @@
 
   CompilationUnitTree getTree();
 
+  /**
+   * @deprecated use reportIssue instead to benefit from precise issue location.
+   */
+  @Deprecated
   void addIssue(Tree tree, JavaCheck check, String message);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/plugins/java/api/JavaFileScannerContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* INFO 2 info

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/plugins/java/api/IssuableSubscriptionVisitor.java [hunk 8]
```diff
@@ -27,10 +27,18 @@
 
 public abstract class IssuableSubscriptionVisitor extends SubscriptionVisitor {
 
+  /**
+   * @deprecated use reportIssue instead to benefit from precise issue location.
+   */
+  @Deprecated
   public void addIssue(Tree tree, String message) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/plugins/java/api/IssuableSubscriptionVisitor.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: java-squid/src/main/java/org/sonar/plugins/java/api/IssuableSubscriptionVisitor.java [hunk 9]
```diff
@@ -27,10 +27,18 @@
...
     context.addIssue(tree, this, message);
   }
 
+  /**
+   * @deprecated use reportIssue instead to benefit from precise issue location.
+   */
+  @Deprecated
   public void addIssue(Tree tree, String message, double effortToFix) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/plugins/java/api/IssuableSubscriptionVisitor.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* INFO 4 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Pull request merged**